session/session.go

```diff
@@ -977,29 +977,42 @@ func (s *session) GetGlobalSysVar(name string) (string, error) {
 }
 
 // SetGlobalSysVar implements GlobalVarAccessor.SetGlobalSysVar interface.
-func (s *session) SetGlobalSysVar(name, value string) error {
-	if name == variable.TiDBSlowLogMasking {
-		name = variable.TiDBRedactLog
+func (s *session) SetGlobalSysVar(name, value string) (err error) {
+	sv := variable.GetSysVar(name)
+	if sv == nil {
+		return variable.ErrUnknownSystemVar.GenWithStackByArgs(name)
+	}
+	if value, err = sv.Validate(s.sessionVars, value, variable.ScopeGlobal); err != nil {
+		return err
+	}
+	if err = sv.SetGlobalFromHook(s.sessionVars, value, false); err != nil {
+		return err
 	}
+
+	return s.updateGlobalSysVar(sv, value)
+}
+
+// SetGlobalSysVarOnly updates the sysvar, but does not call the validation function or update aliases.
+// This is helpful to prevent duplicate warnings being appended from aliases, or recursion.
+func (s *session) SetGlobalSysVarOnly(name, value string) (err error) {
 	sv := variable.GetSysVar(name)
 	if sv == nil {
 		return variable.ErrUnknownSystemVar.GenWithStackByArgs(name)
 	}
-	var sVal string
-	var err error
-	sVal, err = sv.Validate(s.sessionVars, value, variable.ScopeGlobal)
-	if err != nil {
+	if err = sv.SetGlobalFromHook(s.sessionVars, value, true); err != nil {
 		return err
 	}
-	name = strings.ToLower(name)
+	return s.updateGlobalSysVar(sv, value)
+}
+
+func (s *session) updateGlobalSysVar(sv *variable.SysVar, value string) error {
 	// update mysql.tidb if required.
-	if s.varFromTiDBTable(name) {
-		if err = s.setTiDBTableValue(name, sVal); err != nil {
+	if s.varFromTiDBTable(sv.Name) {
```

Comment:
As `sv.Name` yes it will be, since this is directly from the Sysvar definition, which uses the name constants in `sessionctx/variable/tidb_vars.go`.

Reply:
The ToLower function is called when you get the sysvar name, so it is not needed here.
https://github.com/pingcap/tidb/blob/master/sessionctx/variable/sysvar.go#L402-L404

Longer term I would like to change this so that the update for `varFromTiDBTable` is performed in the `SetGlobal` function on the SysVar struct, and the ToLower is not called in this function.